// dom-testing-library/dist/queries.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getAllByTitle = exports.getByTitle = exports.queryAllByTitle = exports.queryByTitle = exports.getAllByTestId = exports.getByTestId = exports.queryAllByTestId = exports.queryByTestId = exports.getAllByAltText = exports.getByAltText = exports.queryAllByAltText = exports.queryByAltText = exports.getAllByLabelText = exports.getByLabelText = exports.queryAllByLabelText = exports.queryByLabelText = exports.getAllByText = exports.getByText = exports.queryAllByText = exports.queryByText = exports.getAllByPlaceholderText = exports.getByPlaceholderText = exports.queryAllByPlaceholderText = exports.queryByPlaceholderText = undefined;

var _matches = require('./matches');

var _getNodeText = require('./get-node-text');

var _prettyDom = require('./pretty-dom');

function debugDOM(htmlElement) {
  return (0, _prettyDom.prettyDOM)(htmlElement, process.env.DEBUG_PRINT_LIMIT || 7000);
}

// Here are the queries for the library.
// The queries here should only be things that are accessible to both users who are using a screen reader
// and those who are not using a screen reader (with the exception of the data-testid attribute query).

function firstResultOrNull(queryFunction) {
  for (var _len = arguments.length, args = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
    args[_key - 1] = arguments[_key];
  }

  var result = queryFunction.apply(undefined, args);
  if (result.length === 0) return null;
  return result[0];
}

function queryAllLabelsByText(container, text) {
  return Array.from(container.querySelectorAll('label')).filter(function (label) {
    return (0, _matches.matches)(label.textContent, label, text);
  });
}

function queryAllByLabelText(container, text) {
  var _ref = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {},
      _ref$selector = _ref.selector,
      selector = _ref$selector === undefined ? '*' : _ref$selector;

  var labels = queryAllLabelsByText(container, text);
  var labelledElements = labels.map(function (label) {
    /* istanbul ignore if */
    if (label.control) {
      // appears to be unsupported in jsdom: https://github.com/jsdom/jsdom/issues/2175
      // but this would be the proper way to do things
      return label.control;
    } else if (label.getAttribute('for')) {
      // we're using this notation because with the # selector we would have to escape special characters e.g. user.name
      // see https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector#Escaping_special_characters
      // <label for="someId">text</label><input id="someId" />
      return container.querySelector(`[id="${label.getAttribute('for')}"]`);
    } else if (label.getAttribute('id')) {
      // <label id="someId">text</label><input aria-labelledby="someId" />
      return container.querySelector(`[aria-labelledby="${label.getAttribute('id')}"]`);
    } else if (label.childNodes.length) {
      // <label>text: <input /></label>
      return label.querySelector(selector);
    } else {
      return null;
    }
  }).filter(function (label) {
    return label !== null;
  }).concat(queryAllByAttribute('aria-label', container, text));

  return labelledElements;
}

function queryByLabelText(container, text, opts) {
  return firstResultOrNull(queryAllByLabelText, container, text, opts);
}

function queryAllByText(container, text) {
  var _ref2 = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {},
      _ref2$selector = _ref2.selector,
      selector = _ref2$selector === undefined ? '*' : _ref2$selector;

  return Array.from(container.querySelectorAll(selector)).filter(function (node) {
    return (0, _matches.matches)((0, _getNodeText.getNodeText)(node), node, text);
  });
}

function queryByText(container, text, opts) {
  return firstResultOrNull(queryAllByText, container, text, opts);
}

var queryAllByTitle = function () {
  for (var _len2 = arguments.length, args = Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
    args[_key2] = arguments[_key2];
  }

  return queryAllByAttribute.apply(undefined, ['title'].concat(args, [{ exact: true }]));
};

var queryByTitle = function () {
  for (var _len3 = arguments.length, args = Array(_len3), _key3 = 0; _key3 < _len3; _key3++) {
    args[_key3] = arguments[_key3];
  }

  return queryByAttribute.apply(undefined, ['title'].concat(args, [{ exact: true }]));
};

function getAllByTitle(container, title) {
  for (var _len4 = arguments.length, rest = Array(_len4 > 2 ? _len4 - 2 : 0), _key4 = 2; _key4 < _len4; _key4++) {
    rest[_key4 - 2] = arguments[_key4];
  }

  var els = queryAllByTitle.apply(undefined, [container, title].concat(rest));
  if (!els.length) {
    throw new Error(`Unable to find an element with the title: ${title}. \n\n${debugDOM(container)}`);
  }
  return els;
}

function getByTitle() {
  for (var _len5 = arguments.length, args = Array(_len5), _key5 = 0; _key5 < _len5; _key5++) {
    args[_key5] = arguments[_key5];
  }

  return firstResultOrNull.apply(undefined, [getAllByTitle].concat(args));
}

// this is just a utility and not an exposed query.
// There are no plans to expose this.
function queryAllByAttribute(attribute, container, text) {
  var _ref3 = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : {},
      _ref3$exact = _ref3.exact,
      exact = _ref3$exact === undefined ? false : _ref3$exact;

  var matcher = exact ? _matches.matchesExact : _matches.matches;
  return Array.from(container.querySelectorAll(`[${attribute}]`)).filter(function (node) {
    return matcher(node.getAttribute(attribute), node, text);
  });
}

// this is just a utility and not an exposed query.
// There are no plans to expose this.
function queryByAttribute() {
  for (var _len6 = arguments.length, args = Array(_len6), _key6 = 0; _key6 < _len6; _key6++) {
    args[_key6] = arguments[_key6];
  }

  return firstResultOrNull.apply(undefined, [queryAllByAttribute].concat(args));
}

var queryByPlaceholderText = queryByAttribute.bind(null, 'placeholder');
var queryAllByPlaceholderText = queryAllByAttribute.bind(null, 'placeholder');
var queryByTestId = function () {
  for (var _len7 = arguments.length, args = Array(_len7), _key7 = 0; _key7 < _len7; _key7++) {
    args[_key7] = arguments[_key7];
  }

  return queryByAttribute.apply(undefined, ['data-testid'].concat(args, [{ exact: true }]));
};
var queryAllByTestId = function () {
  for (var _len8 = arguments.length, args = Array(_len8), _key8 = 0; _key8 < _len8; _key8++) {
    args[_key8] = arguments[_key8];
  }

  return queryAllByAttribute.apply(undefined, ['data-testid'].concat(args, [{ exact: true }]));
};

function queryAllByAltText(container, alt) {
  return Array.from(container.querySelectorAll('img,input,area')).filter(function (node) {
    return (0, _matches.matches)(node.getAttribute('alt'), node, alt);
  });
}

function queryByAltText(container, alt) {
  return firstResultOrNull(queryAllByAltText, container, alt);
}

// getters
// the reason we're not dynamically generating these functions that look so similar:
// 1. The error messages are specific to each one and depend on arguments
// 2. The stack trace will look better because it'll have a helpful method name.

function getAllByTestId(container, id) {
  for (var _len9 = arguments.length, rest = Array(_len9 > 2 ? _len9 - 2 : 0), _key9 = 2; _key9 < _len9; _key9++) {
    rest[_key9 - 2] = arguments[_key9];
  }

  var els = queryAllByTestId.apply(undefined, [container, id].concat(rest));
  if (!els.length) {
    throw new Error(`Unable to find an element by: [data-testid="${id}"] \n\n${debugDOM(container)}`);
  }
  return els;
}

function getByTestId() {
  for (var _len10 = arguments.length, args = Array(_len10), _key10 = 0; _key10 < _len10; _key10++) {
    args[_key10] = arguments[_key10];
  }

  return firstResultOrNull.apply(undefined, [getAllByTestId].concat(args));
}

function getAllByPlaceholderText(container, text) {
  for (var _len11 = arguments.length, rest = Array(_len11 > 2 ? _len11 - 2 : 0), _key11 = 2; _key11 < _len11; _key11++) {
    rest[_key11 - 2] = arguments[_key11];
  }

  var els = queryAllByPlaceholderText.apply(undefined, [container, text].concat(rest));
  if (!els.length) {
    throw new Error(`Unable to find an element with the placeholder text of: ${text} \n\n${debugDOM(container)}`);
  }
  return els;
}

function getByPlaceholderText() {
  for (var _len12 = arguments.length, args = Array(_len12), _key12 = 0; _key12 < _len12; _key12++) {
    args[_key12] = arguments[_key12];
  }

  return firstResultOrNull.apply(undefined, [getAllByPlaceholderText].concat(args));
}

function getAllByLabelText(container, text) {
  for (var _len13 = arguments.length, rest = Array(_len13 > 2 ? _len13 - 2 : 0), _key13 = 2; _key13 < _len13; _key13++) {
    rest[_key13 - 2] = arguments[_key13];
  }

  var els = queryAllByLabelText.apply(undefined, [container, text].concat(rest));
  if (!els.length) {
    var labels = queryAllLabelsByText(container, text);
    if (labels.length) {
      throw new Error(`Found a label with the text of: ${text}, however no form control was found associated to that label. Make sure you're using the "for" attribute or "aria-labelledby" attribute correctly. \n\n${debugDOM(container)}`);
    } else {
      throw new Error(`Unable to find a label with the text of: ${text} \n\n${debugDOM(container)}`);
    }
  }
  return els;
}

function getByLabelText() {
  for (var _len14 = arguments.length, args = Array(_len14), _key14 = 0; _key14 < _len14; _key14++) {
    args[_key14] = arguments[_key14];
  }

  return firstResultOrNull.apply(undefined, [getAllByLabelText].concat(args));
}

function getAllByText(container, text) {
  for (var _len15 = arguments.length, rest = Array(_len15 > 2 ? _len15 - 2 : 0), _key15 = 2; _key15 < _len15; _key15++) {
    rest[_key15 - 2] = arguments[_key15];
  }

  var els = queryAllByText.apply(undefined, [container, text].concat(rest));
  if (!els.length) {
    throw new Error(`Unable to find an element with the text: ${text}. This could be because the text is broken up by multiple elements. In this case, you can provide a function for your text matcher to make your matcher more flexible. \n\n${debugDOM(container)}`);
  }
  return els;
}

function getByText() {
  for (var _len16 = arguments.length, args = Array(_len16), _key16 = 0; _key16 < _len16; _key16++) {
    args[_key16] = arguments[_key16];
  }

  return firstResultOrNull.apply(undefined, [getAllByText].concat(args));
}

function getAllByAltText(container, alt) {
  var els = queryAllByAltText(container, alt);
  if (!els.length) {
    throw new Error(`Unable to find an element with the alt text: ${alt} \n\n${debugDOM(container)}`);
  }
  return els;
}

function getByAltText() {
  for (var _len17 = arguments.length, args = Array(_len17), _key17 = 0; _key17 < _len17; _key17++) {
    args[_key17] = arguments[_key17];
  }

  return firstResultOrNull.apply(undefined, [getAllByAltText].concat(args));
}

exports.queryByPlaceholderText = queryByPlaceholderText;
exports.queryAllByPlaceholderText = queryAllByPlaceholderText;
exports.getByPlaceholderText = getByPlaceholderText;
exports.getAllByPlaceholderText = getAllByPlaceholderText;
exports.queryByText = queryByText;
exports.queryAllByText = queryAllByText;
exports.getByText = getByText;
exports.getAllByText = getAllByText;
exports.queryByLabelText = queryByLabelText;
exports.queryAllByLabelText = queryAllByLabelText;
exports.getByLabelText = getByLabelText;
exports.getAllByLabelText = getAllByLabelText;
exports.queryByAltText = queryByAltText;
exports.queryAllByAltText = queryAllByAltText;
exports.getByAltText = getByAltText;
exports.getAllByAltText = getAllByAltText;
exports.queryByTestId = queryByTestId;
exports.queryAllByTestId = queryAllByTestId;
exports.getByTestId = getByTestId;
exports.getAllByTestId = getAllByTestId;
exports.queryByTitle = queryByTitle;
exports.queryAllByTitle = queryAllByTitle;
exports.getByTitle = getByTitle;
exports.getAllByTitle = getAllByTitle;

/* eslint complexity:["error", 14] */
